etc/systemd/system-generators/zfs-mount-generator.c

```diff
@@ -0,0 +1,1062 @@
+/*
+ * Permission to use, copy, modify, and/or distribute this software for
+ * any purpose with or without fee is hereby granted.
+ *
+ * THE SOFTWARE IS PROVIDED "AS IS" AND THE AUTHOR DISCLAIMS ALL WARRANTIES
+ * WITH REGARD TO THIS SOFTWARE INCLUDING ALL IMPLIED WARRANTIES OF
+ * MERCHANTABILITY AND FITNESS. IN NO EVENT SHALL THE AUTHOR BE LIABLE FOR
+ * ANY SPECIAL, DIRECT, INDIRECT, OR CONSEQUENTIAL DAMAGES OR ANY DAMAGES
+ * WHATSOEVER RESULTING FROM LOSS OF USE, DATA OR PROFITS, WHETHER IN
+ * AN ACTION OF CONTRACT, NEGLIGENCE OR OTHER TORTIOUS ACTION, ARISING OUT
+ * OF OR IN CONNECTION WITH THE USE OR PERFORMANCE OF THIS SOFTWARE.
+ */
+
+
+#include <sys/resource.h>
+#include <sys/types.h>
+#include <sys/time.h>
+#include <sys/stat.h>
+#include <sys/wait.h>
+#include <sys/mman.h>
+#include <semaphore.h>
+#include <stdbool.h>
+#include <unistd.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <time.h>
+#include <regex.h>
+#include <search.h>
+#include <dirent.h>
+#include <string.h>
+#include <stdlib.h>
+#include <limits.h>
+#include <errno.h>
+#include <libzfs.h>
+
+#define	STRCMP ((int(*)(const void *, const void *))&strcmp)
+#define	ABORT_ENOMEM() \
+	do { \
+		fprintf(stderr, PROGNAME ": %d: " \
+		    "not enough memory (L%d)!\n", getpid(), __LINE__); \
+		_exit(1 << 0); \
```

Comment:
Out of curiosity: Why the shift here?

Reply:
Symmetry with all the other returns from the worker process – they're all `(1 << N)` for increasing N, and this is the first one

Reply:
What's the reasoning for different error return codes?

Reply:
Ability to accumulate different error conditions and return them from the main process (for debugging only, these are entirely ignored by systemd during a reload).

Reply:
I see what's going on with the bit flag.
nit: My C foo is very weak (guess why I stuck to shell), but shouldn't this be an enum or a constant or something similar?
Obv not the most important thing in the world.

Reply:
It doesn't really matter, since all faulty exit paths already log the error, and it's just crudely accumulated in the main process. Nothing actually uses these or checks for them explicitly, they're all ORed together and passed to _exit, the only thing they're really useful for is seeing How Many different errors happened at a glance if you echo out `$?` afterward or deconstructing the failure bits if you devnulled stderr, I guess. It's the moral equivalent to returning increasing numbers from `main()`.

Reply:
sure